doc/adr/0003-nats-headers.md

```diff
@@ -0,0 +1,72 @@
+# 3. nats-headers
+
+Date: 2020-05-28
+
+## Status
+
+Accepted
+
+## Context
+
+This document describes NATS Headers from the perspective of clients. NATS headers allow clients to specify additional metadata in the form of headers. The headers borrow strongly from [HTTP Headers](https://tools.ietf.org/html/rfc7230#section-3.2).
+
+The salient points of the HTTP header specification are:
+
+- Each header field consists of a case-insensitive field name followed by a colon (":"), optional leading whitespace, the field value, and optional trailing whitespace.
+- No spaces are allowed between the header field name and colon.
+- Field value may be preceeded or followed by optional whitespace.
+- The specification may allow any number of strange things like comments/tokens etc.
+- The keys can repeat.
+
+The reference NATS Go client repurposes the `http.Header` infrastructure for serializing NATS headers. Other language clients may not be able to reuse header infrastructure outside of the context of HTTP. Furthermore, the implementation of the header library may diverge from what is allowed by the Go `http.Header`. 
```

Comment:
> Furthermore, the implementation of the header library may diverge from what is allowed by the Go `http.Header`

Not sure what you are saying here. If it follows the parser that you just outlined below, how can that differ? And if it does, then assume server accepts and forwards, you can't have a Go client not be able to receive a header sent by say a Node client or vice-versa.

Reply:
Its made to be same as HTTP on purpose.

Reply:
@kozlovic the behavior of the encoder is what the go library does, which may or not be the same in different environments.